### ISSUE TITLE
Repro included: segmentation fault

### ISSUE THREAD

**Issue by metaleap** [ziglang/zig#4367]
`zig version`: `0.5.0+e6a812c82` (Zig's prebuilt-binary distr., approx. 1-2 days old), Void Linux x86_64, kernel 4.19.98_1

You'll get the lone sole message `Segmentation fault` when you `zig run` or `zig test` the below minimal-repro.zig. 

Would have liked to minimize further but it seems like all the still-remaining code-paths in there are somehow prerequisite for the segfault at my end...   =/

The segfault is avoided by removing/commenting line 65 (near the end & marked with `// TODO` comment as line-numbers don't show up here). It builds and prints as expected.

```zig
const std = @import("std");

const TimeInfo = struct {
    start: i64,
    now: ?u64,
};

pub const OptionsRepro = struct {
    err_on_missing_nonvoid_nonoptional_fields: bool = true,
    isStructFieldEmbedded: ?fn (type, []const u8, type) bool,
};

test "" {
    try main();
}

pub fn main() !void {
    var mem = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    // defer mem.deinit();

    var jval = std.json.Value{ .Object = std.json.ObjectMap.init(&mem.allocator) };
    _ = try jval.Object.put("start", std.json.Value{ .Integer = 123 });
    _ = try jval.Object.put("now", std.json.Value{ .Integer = 321 });

    var timeinfo = try unmarshalRepro(TimeInfo, &mem, &jval, &OptionsRepro{
        .isStructFieldEmbedded = defaultIsStructFieldEmbedded,
    });
    std.debug.warn("\n\nYUP:\t{}\n\n", .{timeinfo});
}

pub fn unmarshalRepro(comptime T: type, mem: *std.heap.ArenaAllocator, from: *const std.json.Value, comptime options: *const OptionsRepro) error{
    MissingField,
    UnexpectedInputValueFormat,
    OutOfMemory,
}!T {
    const type_id = comptime @typeId(T);
    const type_info = comptime @typeInfo(T);
    if (type_id == .Int)
        return switch (from.*) {
            .Integer => |jint| @intCast(T, jint),
            else => error.UnexpectedInputValueFormat,
        };
    if (type_id == .Optional) switch (from.*) {
        .Null => return null,
        else => if (unmarshalRepro(type_info.Optional.child, mem, from, options)) |ok|
            return ok
        else |err|
            return err,
    };
    if (type_id == .Struct) {
        switch (from.*) {
            .Object => |*jmap| {
                var ret: T = undefined;
                comptime var i = @memberCount(T);
                inline while (i > 0) {
                    i -= 1;
                    const field_name = @memberName(T, i);
                    const field_type = @memberType(T, i);
                    const field_embed = comptime (@typeId(field_type) == .Struct and options.isStructFieldEmbedded.?(T, field_name, field_type));
                    if (field_embed)
                        @field(ret, field_name) = try unmarshalRepro(field_type, mem, from, options)
                    else if (jmap.getValue(field_name)) |*jval|
                        @field(ret, field_name) = try unmarshalRepro(field_type, mem, jval, options)
                    else if (options.err_on_missing_nonvoid_nonoptional_fields) {
                        return error.MissingField; // TODO! Zig currently segfaults here
                    }
                }
                return ret;
            },
            else => return error.UnexpectedInputValueFormat,
        }
    }
    @compileError(@typeName(T));
}

fn defaultIsStructFieldEmbedded(comptime struct_type: type, field_name: []const u8, comptime field_type: type) bool {
    return false; // std.mem.eql(u8, field_name, @typeName(field_type));
}
```

**Bonus issue!** you can change the segfault into an `Assertion failed`/`This is a bug in the Zig compiler` by changing in line 10 the `isStructFieldEmbedded: ?fn ....` into non-optional, and the call-site in line 59 accordingly.

**Comment by metaleap** [comment]
Should I open a separate issue with the near-identical repro for the above-noted "Assertion failed" variation, or keep them both in (this) one?

**Comment by LemonBoy** [comment]
Same as #4368, this time you need to dump the `ll` output and run it trough `llc` to understand where the problem is.

Long story short, another `nullptr` passes trough and eventually chokes LLVM during the IR verification phase. Here's the culprit:
```llvm
br i1 %59, label %ErrRetReturn2, <null operand!>
```

**Comment by Vexu** [comment]
Fixed by 4618c41fa6ca70f06c7e65762d2f38d57b00818c
